PEP 677: Callable Type Syntax (#2187)
python/peps: pep-XXXX.rst

PEP: 9999
Title: Callable Type Syntax
Author: Steven Troxler <steven.troxler@gmail.com>,
        Pradeep Kumar Srinivasan <gohanpra@gmail.com>
Sponsor: TODO
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: xxx
Python-Version: 3.11
Post-History:

Abstract
========

This PEP introduces a concise and friendly syntax for callable types, supporting the same functionality as ``typing.Callable`` but with an arrow syntax inspired by the syntax for typed function signatures. This allows types like ``Callable[[int, str], bool]`` to be written ``(int, str) -> bool``.

The proposed syntax supports all the functionality provided by ``typing.Callable`` and ``typing.Concatenate``, and is intended to work as a drop-in replacement.

Motivation
==========

> The ``Callable`` type, defined as part of PEP 484, is one of the most commonly used complex types in ``typing`` alongside ``Union`` and collection types like ``Dict`` and ``List``.

Reviewer: ```suggestion
The ``Callable`` type, introduced in PEP 484, is one of the most commonly used complex types in the ``typing`` module alongside ``Union`` and collection types such as ``Dict`` and ``List``.
```

There are four major problems with the existing ``Callable`` type:

> - it is verbose, particularly for more complex function signatures.

Reviewer: ```suggestion
- It is verbose, particularly for more complex function signatures.
```

> - it requires an explicit import, something we no longer require for most of the other

Reviewer: ```suggestion
- It requires an explicit import, something we no longer require for most of the other
```

  very common types after PEP 604 (``|`` for ``Union`` types) and PEP 585 (generic collections)

> - it does not visually represent the way function headers are written.

Reviewer: ```suggestion
- It does not visually represent the way function headers are written.
```

> - it relies on two levels of nested square brackets. This can be quite hard to read,

Reviewer: ```suggestion
- It relies on two levels of nested square brackets. This can be quite hard to read,
```

  especially when the function arguments themselves have square brackets.

> It is common for library authors to make use of untyped or partially-typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. Libraries with less precise types reduce the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.

Reviewer: ```suggestion
It is common for library authors to make use of untyped or partially-typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. If libraries use less precise types, it reduces the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.
```

> It is common for library authors to make use of untyped or partially-typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. Libraries with less precise types reduce the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.

Reviewer: ```suggestion
It is common for library authors to make use of untyped or partially typed callables (e.g. ``Callable[..., Any]``) which we believe is partially a result of the existing types being hard to use. Libraries with less precise types reduce the ability of static analyzers running on downstream projects (including type checkers and security analysis tools) to find problems.
```

> With a succinct, easy-to-use syntax, developers may be less likely to reach for poorly-typed options. Callable types may also be beginner-friendly if we make them look more like function headers, and like the arrow type syntax used by several other popular languages.

Reviewer: ```suggestion
With a succinct, easy-to-use syntax, developers may be less likely to reach for poorly typed options. Callable types may also be beginner-friendly if we make them look more like function headers, and like the arrow type syntax used by several other popular languages.
```

A simplified real-world example from a web server illustrates how the types can be verbose and require many levels of nested square brackets::

    from typing import Awaitable, Callable
    from app_logic import Response, UserSetting

    async def customize_response_for_settings(
        response: Response,
        customizer: Callable[[Response, list[UserSetting]], Awaitable[Response]]
    ) -> Response:
       ...

With our proposal, this code can be abbreviated to::

    from app_logic import Response, UserSetting

    def make_endpoint(
        response: Response,
        customizer: async (Response, list[UserSetting]) -> Response,
    ) -> Response:
        ...

This is shorter and requires fewer imports. It also has far less nesting of square brackets - only one level, as opposed to three in the original code.

Rationale
=========

> The ``Callable`` type is widely used. For example, in typeshed [#typeshed-stats]_ it is the fifth most common complex type, after ``Optional``, ``Tuple``, ``Union``, and ``List``.

Reviewer: Would it be worth noting that this statistic is based on analysis done in October 2021? (Things might always change! The heavy presence of `Optional`, `Tuple`, `Union` and `List` in typeshed is due to a bug in mypy that means PEP 585 and PEP 604 can't yet be used in all situations. But a new release of mypy is just around the corner...)

Author: Good point. I think using current stats is most fair since we're arguing by analogy here, but we should be specific about the time

> The ``Callable`` type is widely used. For example, as of October 2021 it is the fifth most common complex type in typeshed [#typeshed-stats]_, after ``Optional``, ``Tuple``, ``Union``, and ``List``.

Reviewer: ```suggestion
The ``Callable`` type is widely used. For example, as of October 2021 it was the fifth most common complex type in typeshed [#typeshed-stats]_, after ``Optional``, ``Tuple``, ``Union``, and ``List``.
```

> Most of the other commonly used types have had their syntax improved via either PEP 604 or PEP 525. ``Callable`` is used heavily enough to similarly justify a more usable syntax.

Reviewer 2: Should the PEP-525 actually be PEP-585?

Author: Oops, yes!

> We chose to support all the existing semantics of ``typing.Callable``, without adding support for new features. We took this decision after examining how frequently each feature might be used in existing typed and untyped open-source code. We determined that the vast majority of use cases are covered.

Reviewer: ```suggestion
In this proposal, we chose to support all the existing semantics of ``typing.Callable``, without adding support for new features. We took this decision after examining how frequently each feature might be used in existing typed and untyped open-source code. We determined that the vast majority of use cases are covered.
```

We considered adding support for named, optional, and variadic arguments. However, we decided against including these features, as our analysis showed they are infrequently used. When they are really needed, it is possible to type these using Callback Protocols [#callback-protocols]_.

See the Rejected Alternatives section for more detailed discussion about omitted features.

Specification
=============

Typing Behavior
---------------

> Inside of type checkers, the new syntax should be treated with exactly the same semantics as ``typing.Callable``.

Reviewer: ```suggestion
Type checkers should treat the new syntax with exactly the same semantics as ``typing.Callable``.
```

> So a type checker should treat the following pairs exactly the same::

Reviewer: ```suggestion
As such, a type checker should treat the following pairs exactly the same::
```

   from typing import Awaitable, Callable, Concatenate, ParamSpec, TypeVarTuple

    P = ParamSpec("P")
    Ts = TypeVarTuple('Ts')

    f0: (int, str) -> bool
    f0: Callable[[int, str], bool]

    f1: (...) -> bool
    f1: Callable[..., bool]

    f2: async (str) -> str
    f2: Callable[[str], Awaitable[str]]

    f3: (**P) -> bool
    f3: Callable[P, bool]

    f4: (int, **P) -> bool
    f4: Callable[Concatenate[int, P], bool]

    f5: (*Ts) -> bool
    f5: Callable[[*Ts], bool]

    f6: (int, *Ts, str) -> bool
    f6: Callable[[int, *Ts, str], bool]

Grammar and Ast
---------------

> The new syntax we’re proposing can be described by these AST changes ::

Reviewer: ```suggestion
The proposed new syntax can be described by these AST changes ::
```

- The semantics of ``(...) -> bool`` are different from ``(T) -> bool`` for any valid type T: ``(...)`` is a special form indicating ``AnyArguments`` whereas ``T`` is a type parameter in the arguments list.
- ``...`` is used as a placeholder default value to indicate an optional argument in stubs and Callback Protocols. Allowing it in the position of a type could easily lead to confusion and possibly bugs due to typos.

Since ``...`` is meaningless as a type and there are usability concerns, our grammar rules it out and the following is a syntax error::

    (int, ...) -> bool

Incompatibility with other possible uses of ``*`` and ``**``
‘’’’’’’‘’’’’’’‘’’’’’’‘’’’’’’‘‘’’’’’’‘’’’’’’‘’‘’’’’’’‘’’‘’’’’

The use of ``**P`` for supporting PEP 612 ``ParamSpec`` rules out any future proposal using a bare ``**<some_type>`` to type ``kwargs``. This seems acceptable because:
- If we ever do want such a syntax, it would be clearer and to require an argument name anyway so that the type looks more similar to a function signature. In other words, if we ever support typing ``kwargs`` in callable types, we would prefer ``(int, **kwargs: str)`` rather than ``(int, **str)``.

> - PEP 646 unpack syntax would rule out using ``*<some_type>`` for ``args``, and the ``kwargs`` case is similar enough that this rules out a bare ``**<some_type>`` anyway.

Reviewer: ```suggestion
- PEP 646 unpacking syntax would rule out using ``*<some_type>`` for ``args``. The ``kwargs`` case is similar enough that this rules out a bare ``**<some_type>`` anyway.
```

We decided on a simple proposal focused just on improving syntax for the existing ``Callable`` type based on an extensive analysis of existing projects (see [#callable-type-usage-stats]_, [#callback-usage-stats-typed]_, [#callback-usage-stats]_). We determined that the vast majority of callbacks can be correctly described by the existing ``typing.Callable`` semantics:
- Positional parameters: By far the most important case to handle well is simple callable types with positional parameters, such as ``(int, str) -> bool``
- ParamSpec and Concatenate: The next most important feature is good support for PEP 612 ``ParamSpec`` and ``Concatenate`` types like ``(**P) -> bool`` and ``(int, **P) -> bool``. These are common primarily because of the heavy use of decorator patterns in python code.
- TypeVarTuples: The next most important feature, assuming PEP 646 is accepted, is for unpacked types which are common because of cases where a wrapper passes along ``*args`` to some other function.

> Features that other, more complicated proposals would support account for fewer than 2% of the use cases we found. These are already expressible using `Callback Protocols <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_, and since they aren’t common we decided that it made more sense to move forward with a simpler syntax.

Reviewer: ```suggestion
Features that other, more complicated proposals would support account for fewer than 2% of the use cases we found. These are already expressible using `Callback Protocols <https://www.python.org/dev/peps/pep-0544/#callback-protocols>`_, and since they are uncommon we decided that it made more sense to move forward with a simpler syntax.
```

Extended Syntax Supporting Named and Optional Arguments
-------------------------------------------------------

Another alternative was for a compatible but more complex syntax that could express everything in this PEP but also named, optional, and variadic arguments. In this “extended” syntax proposal the following types would have been equivalent::

    class Function(typing.Protocol):
        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
            ...

    Function = (int, y: float, *, z: bool = ..., **kwargs: str) -> bool

Advantages of this syntax include:
- Most of the advantages of the proposal in this PEP (conciseness, PEP 612 support, etc)
- Furthermore, the ability to handle named, optional, and variadic arguments

We decided against proposing it for the following reasons:
- The implementation would have been more difficult, and usage stats demonstrate that fewer than 3% of use cases would benefit from any of the added features.
- The group that debated these proposals was split down the middle about whether these changes are even desirable:

>   - On the one hand they make callable types more expressive. On the other hand, they could easily confuse users who haven’t read the full specification of callable type syntax.

Reviewer: ```suggestion
  - On the one hand, they make callable types more expressive. On the other hand, they could easily confuse users who have not read the full specification of callable type syntax.
```

>   - We believe the simpler syntax proposed in this PEP, which introduces no new semantics and closely mimics syntax in other popular languages like Kotlin, Scala, and TypesScript, are much less likely to confuse users.

Reviewer: ```suggestion
  - We believe the simpler syntax proposed in this PEP, which introduces no new semantics and closely mimics syntax in other popular languages like Kotlin, Scala, and TypesScript, is much less likely to confuse users.
```

> - We intend to implement the current proposal in a way that is forward-compatible with the more complicated extended syntax. So if the community decides after more experience and discussion that we want the additional features they should be straightforward to propose in the future.

Reviewer: ```suggestion
- We intend to implement the current proposal in a way that is forward-compatible with the more complicated extended syntax. If the community decides after more experience and discussion that we want the additional features, they should be straightforward to propose in the future.
```

One alternative we had floated was a syntax much more similar to function signatures.

In this proposal, the following types would have been equivalent::

    class Function(typing.Protocol):
        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
            ...

    Function = (x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool

The benefits of this proposal would have included:
- Perfect syntactic consistency between signatures and callable types.
- Support for more features of function signatures (named, optional, variadic args) that this PEP does not support.

Key downsides that led us to reject the idea include the following:
- A large majority of use cases only use positional-only arguments, and this syntax would be more verbose for that use case, both because of requiring argument names and an explicit ``/``, for example ``(int, /) -> bool`` where our proposal allows ``(int) -> bool``
- The requirement for explicit ``/`` for positional-only arguments has a high risk of causing frequent bugs - which often wouldn’t be detected by unit tests - where library authors would accidentally use types with named arguments.

> - Our analysis suggests that support for ``ParamSpec`` is key, but the scope rules laid out in PEP 612 would have made this difficult.

Reviewer: ```suggestion
- Our analysis suggests that support for ``ParamSpec`` is key, but the scoping rules laid out in PEP 612 would have made this difficult.
```

Other Proposals Considered
--------------------------

> An idea we looked at very early on was to allow using functions as types. This may be a great idea, but we consider less an alternative to better callable types than a major improvement in the usability of Callable Protocols:

Reviewer: It's a little unclear what this "using functions as types" proposal is. Could you add a link to somewhere where it's fleshed out a little more? 🙂

Author: Yep! Unfortunately I don't think there's much to link to, everyone in the typing-sig discussion knows this from some meeings and emails but that's about it. I will see if there's a good email to link to, otherwise we may need to write a small doc on this.

Reviewer: Clearly not "everyone", since I'm on typing-sig 😉

Reviewer: (I think I joined about halfway through the discussion on improved callable syntax)

Author: Makes sense. It's an idea Łukasz Langa raised at the 2021 PyCon typing summit.

The idea is that we could allow writing something like
```
def MyCallbackProtocolType(x: int, /, y: str = ...): ...
```
in place of
```
class MyCallbackProtocolType(Protocol):
    def __call__(self, x: int, /, y: str = ...): ...
```
and then use the function itself as a type representing any callable with a compatible call signature.

Reviewer: Ooh, that would be nice. But yes -- a great idea for a different PEP!

Author: I wrote up a doc on the idea because I don't want to inline too much about this proposal, since it mostly deserves a separate discussion not intermingled with comments about this PEP.
https://docs.google.com/document/d/1rv6CCDnmLIeDrYlXe-QcyT0xNPSYAuO1EBYjU3imU5s/edit

I linked to the doc, and also added a bit more detail directly in the PEP

Reviewer: Ah, that's much better now! And yes, I agree that it's better to keep the details of that proposal out of the main body of this PEP :)

Reviewer: A thought: perhaps this Google Doc could become an appendix to the PEP?

> - Using functions as types wouldn’t give us a new way of describing function types as first class values. Instead, they would require a function definition statement that effectively defines a type alias (much as a Callable Protocol class statement does).

Reviewer: ```suggestion
- Using functions as types would not give us a new way of describing function types as first class values. Instead, they would require a function definition statement that effectively defines a type alias (much as a Callable Protocol class statement does).
```

- Functions-as-types would support almost exactly the same features that Callable Protocols do today: named, optional, and variadic args as well as the ability to define overloads.

> So we think that is an idea for a related PEP, but not a direct substitute for improved Callable syntax.

Reviewer: ```suggestion
Therefore, we think that is an idea for a related PEP, but not a direct substitute for improved Callable syntax.
```

We considered a parentheses-free syntax that would have been even more concise::

    int, str -> bool

We decided against it because this is not visually as similar to existing function header syntax. Moreover, it is visually similar to lambdas, which bind names with no parentheses: ``lambda x, y: x == y``.

> Another idea was a new “special string” syntax an puting the type inside of it, for example ``t”(int, str) -> bool”``. We rejected this because it is not as readable, and it doesn’t seem in line with guidance from the Steering Council on ensuring that type expressions do not diverge from the rest of Python syntax. [#python-types-and-runtime-guidance]_

Reviewer: ```suggestion
Another idea was a new “special string” syntax an puting the type inside of it, for example ``t”(int, str) -> bool”``. We rejected this because it is not as readable, and seems out of step with guidance from the Steering Council on ensuring that type expressions do not diverge from the rest of Python's syntax. [#python-types-and-runtime-guidance]_
```

Backwards Compatibility
=======================

This PEP proposes a major syntax improvement over ``typing.Callable``, but the static semantics are the same.

As such, the only thing we need for backward compatibility is to ensure that types specified via the new syntax behave the same as equivalent ``typing.Callable`` and ``typing.Concatenate`` values they intend to replace.

> There’s no particular interaction between this proposal and ``from __future__ import annotations`` - just like any other type annotation it will be unparsed to a string at module import, and ``typing.get_type_hints`` should correctly evaluate the resulting strings in cases where that is possible.

Reviewer: ```suggestion
There is no particular interaction between this proposal and ``from __future__ import annotations`` - just like any other type annotation it will be unparsed to a string at module import, and ``typing.get_type_hints`` should correctly evaluate the resulting strings in cases where that is possible.
```

This is discussed in more detail in the Runtime Behavior section.

Reference Implementation
========================

We have a working implementation of the AST and Grammar [#callable-type-syntax--shorthand]_ with tests verifying that the grammar proposed here has the desired behaviors.

There is no runtime implementation yet. At a high level we are committed to the following by backward compatibility:
- We will need new object types for both the callable type and concatenation type, tentatively defined in C and exposed as ``types.CallableType`` and ``types.CallableConcatenateType`` in a manner similar to ``types.UnionType``.
- The new types must support existing ``typing.Callable`` and ``typing.Concatenate`` runtime APIs almost exactly:

>   - The ``__repr__`` methods will differ and display the new builtin syntax
>   - But the ``__args__`` and ``__parameters__`` fields must behave the same
>   - And the indexing operation - which returns a new type object with concrete types substituted for various entries in ``__parameters__``, must also be the same.

Reviewer: ```suggestion
  - The ``__repr__`` methods will differ and display the new builtin syntax;
  - But the ``__args__`` and ``__parameters__`` fields must behave the same;
  - And the indexing operation - which returns a new type object with concrete types substituted for various entries in ``__parameters__``, must also be the same.
```

We will return to more details of the runtime behavior, which remain open to discussion other than backward compatibility, in the Open Issues section below.

Open Issues
===========

Details of the Runtime API
--------------------------

The new runtime objects to which this syntax evaluates will remain backward-compatible with the ``typing.Callable`` and ``typing.Concatenate`` types they replace, other than details like ``__repr__`` where some behavior change makes sense.

But we also believe that we should have a new runtime API with more structured data access, since:

> - Callable types have a more complicated shape than other generics, especially given the behavior when using ``...`` and ``typing.Concatenate``

Reviewer: ```suggestion
- Callable types have a more complicated shape than other generics, especially given the behavior when using ``...`` and ``typing.Concatenate``.
```

- In the future we might want to add more features, such as support for named and optional arguments, that would be even more difficult to describe well using only ``__args__`` and ``__parameters___``.

Our tentative plan is to define enough new builtins for the runtime data to mirror the shape of the AST, but other options are also possible. See [#runtime-behavior-specification]_ for a detailed description of the current plan and a place to discuss other ideas.

Once the runtime behavior is fully defined we will add a complete evaluation model and description of behavior to this PEP.

Optimizing ``SyntaxError`` messages
-----------------------------------

The current reference implementation has a fully-functional parser and all edge cases presented here have been tested.

> But there are some known cases where the errors are not as informative as we would like. For example, because ``(int, ...) -> bool`` is illegal but ``(int, ...)`` is a valid tuple, we currently produce a syntax error flagging the ``->`` as the problem even though the real cause of the error is using ``...`` as an argument type.

Reviewer: ```suggestion
However, there are some known cases where the errors are not as informative as we would like. For example, because ``(int, ...) -> bool`` is illegal but ``(int, ...)`` is a valid tuple, we currently produce a syntax error flagging the ``->`` as the problem even though the real cause of the error is using ``...`` as an argument type.
```

Author: I think I'll skip this one for now, I had it drilled into me in high school that "But" is okay and "However" is bad, can't overcome that training :)

> This is not part of the specification per se but is an important detail to add  ress in our implementation. The solution will likely involve adding ``invalid_.*`` rules to ``python.gram`` and customizing error messages.

Reviewer: ```suggestion
This is not part of the specification *per se* but is an important detail to address in our implementation. The solution will likely involve adding ``invalid_.*`` rules to ``python.gram`` and customizing error messages.
```

Resources
=========

Background and History
----------------------

PEP 484 [#pep-484-function-type-hints]_ specifies a very similar syntax for function type hint *comments* for use in code that needs to work on Python 2.7, for example::

    def f(x, y):
        # type: (int, str) -> bool
        ...

> At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types, but we have since begun to do so, e.g. with PEP 604

Reviewer: ```suggestion
At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types. However, we have since begun to do so, e.g. with PEP 604.
```

PEP: 9999
Title: Callable Type Syntax
Author: Steven Troxler <steven.troxler@gmail.com>,
        Pradeep Kumar Srinivasan <gohanpra@gmail.com>
Sponsor: TODO
Status: Draft
Type: Standards Track
Content-Type: text/x-rst
Created: xxx
Python-Version: 3.11
Post-History:

Abstract
========

This PEP introduces a concise and friendly syntax for callable types, supporting the same functionality as ``typing.Callable`` but with an arrow syntax inspired by the syntax for typed function signatures. This allows types like ``Callable[[int, str], bool]`` to be written ``(int, str) -> bool``.

> The proposed syntax supports all the functionality provided by ``typing.Callable`` and ``typing.Concatenate`` and could eventually replace them entirely.

Author: Yeah I think this needs work, thanks for pointing it out.

I don't think we have consensus about whether to discuss deprecation. I'd rather not bring it up at all because that's not conventional (PEP 604 aims to deprecate Union but does not say so).

> The proposed syntax supports all the functionality provided by ``typing.Callable`` and ``typing.Concatenate`` and could eventually replace them entirely.

Author: For now I opted to go with "drop-in replacement" language, which I think makes the point we want without hinting at deprecation. We'll see what other reviewers think :)

> The proposed syntax supports all the functionality provided by ``typing.Callable`` and ``typing.Concatenate`` and could eventually replace them entirely.

Reviewer: Nice, I think the revised wording is great!

A simplified real-world example from a web server illustrates how the types can be verbose and require many levels of nested square brackets::

    from typing import Callable
    from app_logic import Response, UserSetting

    async def customize_response_for_settings(
        response: Response,
        customizer: Callable[[Response, list[UserSetting]], Awaitable[Response]]
    ) -> Response:
       ...

With our proposal, this code can be abbreviated to::

    from app_logic import Response, UserSetting

    def make_endpoint(
        response: Response,
        customizer: async (Response, list[UserSetting]) -> Response,
    ) -> Response:
        ...

This is shorter and requires fewer imports. It also has far less nesting of square brackets - only one level, as opposed to three in the original code.

Rationale
=========

The ``Callable`` type is widely used. For example, in typeshed [#typeshed-stats]_ it is the fifth most common complex type, after ``Optional``, ``Tuple``, ``Union``, and ``List``.

> Most of the other commonly used types have gotten improved syntax either via PEP 604 or PEP 525.``Callable`` is used heavily enough to similarly justify a more usable syntax.

Author: I don't mind this change, but I'm not sure it's necessary.

Other PEPs seem to be mixed in whether they just use PEP XXX versus :pep:`XXX` and it appears to format the same way on the website regardless.

> Most of the other commonly used types have gotten improved syntax either via PEP 604 or PEP 525.``Callable`` is used heavily enough to similarly justify a more usable syntax.

Reviewer: Okay fair! I was just applying standard rst rules from other parts of Python docs -- didn't realise PEPs worked differently 🙂

> Most of the other commonly used types have gotten improved syntax either via PEP 604 or PEP 525.``Callable`` is used heavily enough to similarly justify a more usable syntax.

Reviewer: With this specific instance, though, I do also feel like "have gotten improved syntax" is slightly inelegant writing 🙂

> Most of the other commonly used types have gotten improved syntax either via PEP 604 or PEP 525.``Callable`` is used heavily enough to similarly justify a more usable syntax.

Author: Oh oops, I missed that part of the line.

I'l apply the wording change, and wait for more guidance with respect to the PEP references.

Thanks!

One alternative we had floated was a syntax much more similar to function signatures.

In this proposal, the following types would have been equivalent::

    class Function(typing.Protocol):
        def f(self, x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool:
            ...

    Function = (x: int, /, y: float, *, z: bool = ..., **kwargs: str) -> bool

The benefits of this proposal would have included
- Perfect syntactic consistency between signatures and callable types.
- Support for more features of function signatures (named, optional, variadic args) that this PEP does not support.

Key downsides that led us to reject the idea include the following:
- A large majority of use cases only use positional-only arguments, and this syntax would be more verbose for that use case, both because of requiring argument names and an explicit ``/``, for example ``(int, /) -> bool`` where our proposal allows ``(int) -> bool``
- The requirement for explicit ``/`` for positional-only arguments has a high risk of causing frequent bugs - which often wouldn’t be detected by unit tests - where library authors would accidentally use types with named arguments.

> - Our analysis suggests that support for ``ParamSpec`` is key, but the scope rules laid out in PEP 612 would have made this difficult.

Reviewer: ```suggestion
- Our analysis suggests that support for ``ParamSpec`` is key, but the scope rules laid out in :pep:`612` would have made this difficult.
```

As such, the only thing we need for backward compatibility is to ensure that types specified via the new syntax behave the same as equivalent ``typing.Callable`` and ``typing.Concatenate`` values they intend to replace.

There’s no particular interaction between this proposal and ``from __future__ import annotations`` - just like any other type annotation it will be unparsed to a string at module import, and ``typing.get_type_hints`` should correctly evaluate the resulting strings in cases where that is possible.

This is discussed in more detail in the Runtime Behavior section.

Reference Implementation
========================

We have a working implementation of the AST and Grammar [#callable-type-syntax--shorthand]_ with tests verifying that the grammar proposed here has the desired behaviors.

There is no runtime implementation yet. At a high level we are committed to the following by backward compatibility:
- We will need new object types for both the callable type and concatenation type, tentatively defined in C and exposed as ``types.CallableType`` and ``types.CallableConcatenateType`` in a manner similar to ``types.UnionType``.

> - The new types must support existing ``typing.Callable`` and ``typing.Concatenate`` runtime apis almost exactly

Reviewer: ```suggestion
- The new types must support existing ``typing.Callable`` and ``typing.Concatenate`` runtime APIs almost exactly:
```

Once the runtime behavior is fully defined we will add a complete evaluation model and description of behavior to this PEP.

Optimizing ``SyntaxError`` messages
-----------------------------------

The current reference implementation has a fully-functional parser and all edge cases presented here have been tested.

But there are some known cases where the errors are not as informative as we would like. For example, because ``(int, ...) -> bool`` is illegal but ``(int, ...)`` is a valid tuple, we currently produce a syntax error flagging the ``->`` as the problem even though the real cause of the error is using ``...`` as an argument type.

This is not part of the specification per se but is an important detail to add  ress in our implementation. The solution will likely involve adding ``invalid_.*`` rules to ``python.gram`` and customizing error messages.

Resources
=========

Background and History
----------------------

PEP 484 [#pep-484-function-type-hints]_ specifies a very similar syntax for function type hint *comments* for use in code that needs to work on Python 2.7, for example::

    def f(x, y):
        # type: (int, str) -> bool
        ...

> At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types, but we have since begun to do so, e.g. with PEP 604

Reviewer: ```suggestion
At that time we used indexing operations to specify generic types like ``typing.Callable`` because we decided not to add syntax for types, but we have since begun to do so, e.g. with :pep:`604`.
```

        def wrapper(retry_once: bool, *args: P.args, **kwargs: P.kwargs):
            ...
        return wrapper

    ...

With these changes, the incorrect attempt to pass ``z`` to ``f``
produces a typecheck error as we would like.

Four usability problems with the way ``typing.Callable`` is
represented may explain why library authors often do not use its full
power:

- It is verbose, particularly for more complex function signatures.
- It does not visually represent the way function headers are written,
  which can make it harder to learn and use.
- It requires an explicit import, something we no longer require for
  most of the other very common types after PEP 604 (``|`` for
  ``Union`` types) and PEP 585 (generic collections)
- It relies on two levels of nested square brackets. This can be quite
  hard to read, especially when the function arguments themselves have
  square brackets.

With our proposed syntax, the properly-typed decorator example becomes
concise and the type representations are visually descriptive::

    from typing import Any, ParamSpec, TypeVar

    V = TypeVar("R")
    P = ParamSpec("P")

>     def with_retries(f: (**P) -> R) -> (bool, **P) -> R:

Reviewer: I'm generally extremely supportive of this PEP, but I worry about this example. The number of arrows in one signature makes my eyes glaze over. The thing is, I'm 90% sure that this will be vastly improved by appropriate syntax highlighting in an IDE — but maybe that should be mentioned?

Reviewer 2: Also simple whitespace trick might make it more readable:
```
def with_retries(
    f: (**P) -> R,
) -> (bool, **P) -> R:
```

Reviewer 2: Can PEP document source use the `.. code-block:: python` RST directive to make the examples use the Python syntax highlighting?

Reviewer: > Can PEP document source use the `.. code-block:: python` RST directive to make the examples use the Python syntax highlighting?

Not currently, I don't think, sadly — though there's a proposal to rework the way they're rendered that would give us syntax highlighting 🙏 (though I'm not sure we'd gain much from that with this PEP, anyway, since we're proposing new syntax?) https://discuss.python.org/t/request-for-comment-making-pep-rendering-self-contained/10774

Author: It's probably not the best intro example but my creativity gave out and the easiest realistic use case I could think of was a decorator.

We'll see if @pradeep90 can come up with a better one. The problem is that we're supposed to be highlighting the kind of situation where people give up and use partially-typed functions, which means the example has to be a little uglier than the basic intro-to-callable-type.

that it is common for library authors to make use of untyped or
partially-typed callables (e.g. ``Callable[..., Any]`` or a bare
``Callable``) which we believe is partially a result of the existing
types being hard to use. But using this kind of partially-typed
callable can negate the benefits of static typing. For example,
consider the following code::

    from typing import Any, Callable

    def with_retries(f: Callable[..., Any]) -> Callable[..., Any]:
        def wrapper(retry_once, *args, **kwargs):
            if retry_once:
                try: return f(*args, **kwargs)
                except Exception: pass
            return f(*args, **kwargs)
        return wrapper

    @with_retries
    def f(x: int) -> int:
        return x

    f(True, z=15)

The decorator above is clearly not intended to modify the type of the
function it wraps, but because it uses ``Callable[..., Any]`` it
accidentally eliminates the annotations on ``f``, and type checkers
will accept the code above even though it is sure to crash at
runtime. A correct version of this code would look like this::

    from typing import Any, Callable, Concatenate, ParamSpec, TypeVar

>     V = TypeVar("R")

Reviewer 2: ```suggestion
    R = TypeVar("R")
```

    P = ParamSpec("P")

    def with_retries(f: Callable[P, R]) -> Callable[Concatenate[bool, P] R]:
        def wrapper(retry_once: bool, *args: P.args, **kwargs: P.kwargs):
            ...
        return wrapper

    ...

With these changes, the incorrect attempt to pass ``z`` to ``f``
produces a typecheck error as we would like.

Four usability problems with the way ``typing.Callable`` is
represented may explain why library authors often do not use its full
power:

- It is verbose, particularly for more complex function signatures.
- It does not visually represent the way function headers are written,
  which can make it harder to learn and use.
- It requires an explicit import, something we no longer require for
  most of the other very common types after PEP 604 (``|`` for
  ``Union`` types) and PEP 585 (generic collections)
- It relies on two levels of nested square brackets. This can be quite
  hard to read, especially when the function arguments themselves have
  square brackets.

With our proposed syntax, the properly-typed decorator example becomes
concise and the type representations are visually descriptive::

    from typing import Any, ParamSpec, TypeVar

>     V = TypeVar("R")

Reviewer 2: ```suggestion
    R = TypeVar("R")
```

> PEP: 9999

Reviewer 3: You got an extra stray file

Author: Yep, just realized that and removed it. Thanks!
